File: C++/RegEx.cpp
#include <iostream>
#include <regex>

using namespace std;

void RegEx(){

	cout <<
	endl <<
	"これから正規表現を試します" << endl;

	string text="The Quick Brown Fox Jumps Over The Lazy Dog";
	cout <<
		text << " →" << endl;

	cout <<
	endl <<

	"検索" << endl;
		smatch found;
		/*
			stringに対する検索: smatch
			char*に対する検索: cmatch
		*/
		if (regex_search(text,found,regex("the ([a-z]+) ([a-z]+) fox",regex_constants::icase))) {
			cout <<
			"マッチ:" << found.str() << " (" << found.position() << "から" << found.length() << "文字)" << endl <<
			"サブマッチ:" << endl;
			for (int n=0;n<3;n++) {
				cout << found[n].str() << " (" << found[n].length() << "文字)" << endl;
			}
		}
		else cout << "マッチしていません";

	cout <<
	endl <<

	"マッチの確認" << endl;
		regex re=regex("fox jumps",regex::icase);
		// 部分一致を含む
		if (regex_search(text,re)) cout << "狐が飛んでいます" << endl;
		else cout << "狐は飛んでいません" << endl;
		// 完全一致のみ検索
		if (regex_match(text,re)) cout << "ただそれだけです" << endl;
		else cout << "言いたいことは他にもある" << endl;
		/*
			regex_searchは部分一致で true/false
			regex_matchは完全一致で true/false
		*/

	cout <<
	endl <<

	"置換" << endl;
		string replaced=regex_replace(text,regex("([a-z]+)o([a-z]+)"),"« $1ö$2 »");
		cout <<
		replaced << endl;
		replaced=regex_replace(text,regex("([a-z]+)o([a-z]+)"),"« $1ö$2 »",regex_constants::format_first_only);
		cout <<
		"format_first_only:" << endl <<
		replaced << endl;

}

File: Java/Legacy/RegEx.java
import java.util.regex.*;

public class RegEx {

	private static String text="The Quick Brown Fox Jumps Over The Lazy Dog";

	public static void main(String[] args) {

		Utility.println(
			"",
			"これから正規表現を試します",

			"",

			text+" → ",
			""
		);

		Utility.println("検索");
		Matcher m=Pattern.compile("(?i)the ([a-z]+) ([a-z]+) fox").matcher(text);
		if (m.find()) {
			Utility.println(m.group()+" ("+m.start()+"~"+m.end()+")");
		}
		else Utility.println("マッチしていません");

		Utility.pnl(1);

		Utility.println("マッチの確認");
		if (text.matches("(?i).*fox jumps.*")) Utility.println("狐が飛んでいます");
		else Utility.println("狐は飛んでいません");
			// string.matchesは完全一致であることに注意 (前後に .* を付加して部分一致にする)
			// (?i)を正規表現中に埋め込むことで,ignoreCaseにできる

		Pattern p=Pattern.compile("(?i)dog jumps");
		if (p.matcher(text).find()) Utility.println("犬が飛んでいます");
		else Utility.println("犬は飛んでいません");

		Utility.pnl(1);

		Utility.println("置換");
		// 正規表現が使えない置換
		String replaced=text.replace("Jump","Skip");
		Utility.println(
			"string.replace:",
			replaced
		);
		// 正規表現が使える置換
		replaced=text.replaceAll("(?i)([a-z]+)o([a-z]+)","« $1ö$2 »");
		Utility.println(
			"string.replaceAll:",
			replaced
		);
		replaced=text.replaceFirst("(?i)([a-z]+)o([a-z]+)","« $1ö$2 »");
		Utility.println(
			"string.replaceFirst:",
			replaced
		);

		Utility.pnl(1);

		Utility.println("分割と結合");
		String[] array=text.split("(the|s)? ");
		String arranged=String.join("_",array);
		Utility.println(arranged);

		Utility.pnl(2);

	}

}